### PR TITLE
Add snapshot feature to chain command

### PR DESCRIPTION
**Pull request by area363**
This PR adds the `snapshot` feature to the `headless chain command`, mostly taken from https://github.com/planetarium/NineChronicles.Snapshot. This should make maintenance much easier.

### CODE REVIEW

File: NineChronicles.Headless.Executable.Tests/Commands/ChainCommandTest.cs [hunk 1]
```diff
@@ -220,6 +220,97 @@ public void PruneState(StoreType storeType)
...
+            chain.ExecuteActions(chain.Tip);
+            var indexCountBeforeSnapshot = store.CountIndex(chainId);
+            const int blockEpochUnitSeconds = 86400;
+            var blockEpoch = (int)(chain.Tip.Timestamp.ToUnixTimeSeconds() / blockEpochUnitSeconds);
+            var genesisBlockEpoch = blockEpoch - 1;
+            var genesisHash = chain.Genesis.Hash;
+            store.Dispose();
+            stateStore.Dispose();
+            const string apv = "1";
+            var outputDirectory = Path.Combine(Path.GetTempPath(), "test");
```
Comment (moreal) [inline NineChronicles.Headless.Executable.Tests/Commands/ChainCommandTest.cs:271]:
```suggestion
            var outputDirectory = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
```

Comment (moreal) [inline NineChronicles.Headless.Executable.Tests/Commands/ChainCommandTest.cs:271]:
```suggestion
            var outputDirectory = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
```

File: NineChronicles.Headless.Executable.Tests/Commands/ChainCommandTest.cs [hunk 2]
```diff
@@ -220,6 +220,97 @@ public void PruneState(StoreType storeType)
...
+                outputDirectory,
+                _storePath,
+                1,
+                ChainCommand.SnapshotType.All);
+            IStore storeAfterSnapshot = storeType.CreateStore(_storePath);
+            chainId = storeAfterSnapshot.GetCanonicalChainId() ?? new Guid();
+            var tipHashAfterSnapshot = storeAfterSnapshot.IndexBlockHash(chainId, -1);
+            var expectedGenesisPartitionSnapshotPath = Path.Combine(Path.GetTempPath(), "test", "partition", $"snapshot-{genesisBlockEpoch}-{genesisBlockEpoch}.zip");
+            var expectedGenesisMetadataPath = Path.Combine(Path.GetTempPath(), "test", "metadata", $"snapshot-{genesisBlockEpoch}-{genesisBlockEpoch}.json");
+            var expectedFullSnapshotPath = Path.Combine(Path.GetTempPath(), "test", "full", $"{genesisHash}-snapshot-{tipHashAfterSnapshot}.zip");
```
Comment (moreal) [inline NineChronicles.Headless.Executable.Tests/Commands/ChainCommandTest.cs:284]:
```suggestion
            var expectedGenesisPartitionSnapshotPath = Path.Combine(outputDirectory, "partition", $"snapshot-{genesisBlockEpoch}-{genesisBlockEpoch}.zip");
            var expectedGenesisMetadataPath = Path.Combine(outputDirectory, "metadata", $"snapshot-{genesisBlockEpoch}-{genesisBlockEpoch}.json");
            var expectedFullSnapshotPath = Path.Combine(outputDirectory, "full", $"{genesisHash}-snapshot-{tipHashAfterSnapshot}.zip");
```

File: NineChronicles.Headless.Executable.Tests/Commands/ChainCommandTest.cs [hunk 3]
```diff
@@ -220,6 +220,97 @@ public void PruneState(StoreType storeType)
...
+
+            _command.Snapshot(
+                apv,
+                outputDirectory,
+                _storePath,
+                1,
+                ChainCommand.SnapshotType.All);
+            var expectedPartitionSnapshotPath = Path.Combine(Path.GetTempPath(), "test", "partition", $"snapshot-{blockEpoch}-{blockEpoch}.zip");
+            var expectedStateSnapshotPath = Path.Combine(Path.GetTempPath(), "test", "state", "state_latest.zip");
+            var expectedMetadataPath = Path.Combine(Path.GetTempPath(), "test", "metadata", $"snapshot-{blockEpoch}-{blockEpoch}.json");
```
Comment (moreal) [inline NineChronicles.Headless.Executable.Tests/Commands/ChainCommandTest.cs:299]:
```suggestion
            var expectedPartitionSnapshotPath = Path.Combine(outputDirectory, "partition", $"snapshot-{blockEpoch}-{blockEpoch}.zip");
            var expectedStateSnapshotPath = Path.Combine(outputDirectory, "state", "state_latest.zip");
            var expectedMetadataPath = Path.Combine(outputDirectory, "metadata", $"snapshot-{blockEpoch}-{blockEpoch}.json");
```

Comment (moreal) [inline NineChronicles.Headless.Executable.Tests/Commands/ChainCommandTest.cs:299]:
```suggestion
            var expectedPartitionSnapshotPath = Path.Combine(outputDirectory, "partition", $"snapshot-{blockEpoch}-{blockEpoch}.zip");
            var expectedStateSnapshotPath = Path.Combine(outputDirectory, "state", "state_latest.zip");
            var expectedMetadataPath = Path.Combine(outputDirectory, "metadata", $"snapshot-{blockEpoch}-{blockEpoch}.json");
```

**Review by moreal — APPROVED** [review]

**Review by U-lis — APPROVED** [review]

**Comment by area363** [comment]
@moreal Thanks for the feedback! I'll apply them and request another review 🙇

**Review by moreal — APPROVED** [review]

**Pull request merged**